.github/workflows/build-gem.yml

```diff
@@ -0,0 +1,165 @@
+name: Build gem
+
+on:
+  workflow_dispatch:
+    inputs:
+      push:
+        description: Push gem
+        required: true
+        type: boolean
+        default: true
+  push:
+    branches:
+      - "**"
+
+env:
+  GEM_HOST: 'https://rubygems.pkg.github.com/DataDog'
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        type:
+          - final
+          - dev
+    runs-on: ubuntu-latest
+    name: Build gem (${{ matrix.type }})
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - uses: ruby/setup-ruby@31a7f6d628878b80bc63375a93ae079ec50a1601 # v1.143.0
+        with:
+          ruby-version: '3.2'
+          bundler-cache: true # runs 'bundle install' and caches installed gems automatically
+      - name: Patch version
+        if: ${{ matrix.type != 'final' }}
+        run: |
+          # Obtain context information
+          git_ref='${{ github.ref }}'
+          git_branch="$(echo "${git_ref}" | sed -e 's#^refs/heads/##')"
+          git_sha='${{ github.sha }}'
+          gha_run_id='${{ github.run_id }}'
+
+          # Output info for CI debug
+          echo git_ref="${git_ref}"
+          echo git_branch="${git_branch}"
+          echo git_sha="${git_sha}"
+          echo gha_run_id="${gha_run_id}"
+
+          # Sanitize for ruby version usage
+          git_branch_sanitized="$(echo "$git_branch" | sed -e 's/[^a-zA-Z0-9+]\{1,\}/./g')"
+          echo git_branch_sanitized="${git_branch_sanitized}"
+
+          # Shorten commit sha
+          git_sha_short="${git_sha:0:12}"
+          echo git_sha_short="${git_sha_short}"
+
+          # Set component values:
+          # - PRE is `dev` to denote being a development version and
+          #   act as a categorizer.
+          # - BUILD starts with CI run id for ordering.
+          # - BUILD has CI run id for traceability, prefixed by `gha`
+          #   for identification.
+          # - BUILD has commit next for traceability, prefixed git-describe
+          #   style by `g` for identification.
+          # - BUILD has branch name last since it has to be separated
+          #   by dots and thus has variable version segment size and
+          #   unpredictable ordering; it can thus be reliably extracted
+          #   and does not impair readability in lists
+          PRE='${{ matrix.type }}'
+          BUILD="gha${gha_run_id}.g${git_sha_short}.${git_branch_sanitized}"
```

Comment:
Minor, looking at your comment in another thread:

> - `1.10.1.alpha.gha4745620001.gfe6c0807f40b.foo.bar.branch.name`
> - `1.10.1.ci.gha4745620001.gfe6c0807f40b.foo.bar.branch.name`
> - `1.10.1.dev.gha4745620001.gfe6c0807f40b.foo.bar.branch.name`

Is it for me that the `g` in front of the sha hash a bit potentially confusing since it looks like part of the hash? (I know the hash is hex but lol it still looks like part of it visually, or if you double-click to select it)

So I'd suggest removing it maybe, but up to you :)

Reply:
As I mentioned, this is inspired by the format of `git describe`:

```
$ git describe --tags
v1.11.0.beta1-9-g2fcff4acc
```

The `g` prefix denotes that this is a git sha, so that it can readily be identified among other pieces of data that could look like a git sha (e.g I was suggesting hashing the branch name).

Note: I prefixed the run id with `gha` for similar reasons.

Reply:
It's not particularly obvious to me, but since we need to pick something, let's go with your approach ;)